### PR TITLE
Bugfix/null player state

### PR DESCRIPTION
**Pull request by BenNaccarato**
#### Description
A bug fix for the null player state issue. This was due to a lack of authority when resolving missing component. This resulted in some null-refs in the PlayerState. The fix was to store a list of of updates that failed to send due to a lack of authority, and process them once authority is gained.

#### Tests
Tested in ShooterGame

#### Primary reviewers
@m-samiec @Vatyx

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 1]
```diff
@@ -163,6 +163,10 @@ void USpatialReceiver::UpdateShadowData(Worker_EntityId EntityId)
 
 void USpatialReceiver::OnAuthorityChange(Worker_AuthorityChangeOp& Op)
 {
+	if (Op.authority == 1)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]:
```suggestion
	if (Op.authority == WORKER_AUTHORITY_AUTHORITATIVE)
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp [hunk 2]
```diff
@@ -276,6 +276,13 @@ void USpatialSender::SendComponentUpdates(UObject* Object, FClassInfo& Info, USp
 		if (!NetDriver->StaticComponentView->HasAuthority(EntityId, Update.component_id))
 		{
 			UE_LOG(LogSpatialSender, Warning, TEXT("Trying to send component update but don't have authority! Update will not be sent. Component Id: %d, entity: %lld"), Update.component_id, EntityId);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]:
```suggestion
			UE_LOG(LogSpatialSender, Warning, TEXT("Trying to send component update but don't have authority! Update will be queued and sent when authority gained. Component Id: %d, entity: %lld"), Update.component_id, EntityId);
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp [hunk 3]
```diff
@@ -276,6 +276,13 @@ void USpatialSender::SendComponentUpdates(UObject* Object, FClassInfo& Info, USp
...
+			UE_LOG(LogSpatialSender, Warning, TEXT("Queueing the update, to send upon receiving authority."));
+			
+			// This is a temporary fix. A task improve this has been created: https://improbableio.atlassian.net/browse/UNR-955
+			// It may be the case that upon resolving a component, we do not have authority to send the update. In this case, we queue the update, to send upon receiving authority.
+			// Note: This will break in a multi-worker context, if we try to create an entity that we don't intend to have authority over. For this reason, this fix is only temporary.
+			TArray<Worker_ComponentUpdate>* UpdatesQueuedUntilAuthority = &UpdatesQueuedUntilAuthorityMap.FindOrAdd(EntityId);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]:
```suggestion
			TArray<Worker_ComponentUpdate>& UpdatesQueuedUntilAuthority = UpdatesQueuedUntilAuthorityMap.FindOrAdd(EntityId);
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp [hunk 4]
```diff
@@ -276,6 +276,13 @@ void USpatialSender::SendComponentUpdates(UObject* Object, FClassInfo& Info, USp
...
+			UpdatesQueuedUntilAuthority->Add(Update);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]:
```suggestion
			UpdatesQueuedUntilAuthority.Add(Update);
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp [hunk 5]
```diff
@@ -284,6 +291,20 @@ void USpatialSender::SendComponentUpdates(UObject* Object, FClassInfo& Info, USp
 }
 
 
+// Apply (and clean up) any updates queued, due to being sent previously when they didn't have authority.
+void USpatialSender::ProcessUpdatesQueuedUntilAuthority(Worker_EntityId EntityId)
+{
+	if (UpdatesQueuedUntilAuthorityMap.Contains(EntityId))
+	{
+		TArray<Worker_ComponentUpdate> UpdatesQueuedUntilAuthority = UpdatesQueuedUntilAuthorityMap[EntityId];
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]: Making a copy here. Take it by pointer instead.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp [hunk 6]
```diff
@@ -276,6 +276,13 @@ void USpatialSender::SendComponentUpdates(UObject* Object, FClassInfo& Info, USp
 		if (!NetDriver->StaticComponentView->HasAuthority(EntityId, Update.component_id))
 		{
 			UE_LOG(LogSpatialSender, Warning, TEXT("Trying to send component update but don't have authority! Update will not be sent. Component Id: %d, entity: %lld"), Update.component_id, EntityId);
+			UE_LOG(LogSpatialSender, Warning, TEXT("Queueing the update, to send upon receiving authority."));
```
Comment (girayimprobable) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]: Let's either remove this line or reduce it to `Log`. Nothing is "wrong" so shouldn't be a warning, and some of our customers will want to run at 0 warnings.

If you keep it, add a bit more info like component id, entity etc.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp [hunk 7]
```diff
@@ -276,6 +276,13 @@ void USpatialSender::SendComponentUpdates(UObject* Object, FClassInfo& Info, USp
...
+			
+			// This is a temporary fix. A task improve this has been created: https://improbableio.atlassian.net/browse/UNR-955
```
Comment (girayimprobable) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]: Let's just refer to it as UNR-955.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp (hunk 2 above)
Comment (girayimprobable) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]: This should be a `Log` as well probably

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp [hunk 8]
```diff
@@ -275,7 +275,13 @@ void USpatialSender::SendComponentUpdates(UObject* Object, FClassInfo& Info, USp
 	{
 		if (!NetDriver->StaticComponentView->HasAuthority(EntityId, Update.component_id))
 		{
-			UE_LOG(LogSpatialSender, Warning, TEXT("Trying to send component update but don't have authority! Update will not be sent. Component Id: %d, entity: %lld"), Update.component_id, EntityId);
+			UE_LOG(LogSpatialSender, Log, TEXT("Trying to send component update but don't have authority! Update will be queued and sent when authority gained. Component Id: %d, entity: %lld"), Update.component_id, EntityId);
+
+			// This is a temporary fix. A task improve this has been created: UNR-955
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]:
```suggestion
			// This is a temporary fix. A task to improve this has been created: UNR-955
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 9]
```diff
@@ -189,6 +189,12 @@ void USpatialReceiver::HandleActorAuthority(Worker_AuthorityChangeOp& Op)
 			return;
 		}
 
+		// TODO UNR-955 - Remove this once batch EntityIds are in.
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]:
```suggestion
		// TODO UNR-955 - Remove this once batch reservation of EntityIds are in.
```

**Pull request merged**